### ISSUE TITLE
Convert whole and fraction to strings

### ISSUE THREAD

**Issue by webmaster128** [iov-one/iov-core#502]

**Comment by ethanfrey** [comment]
Fixed point format:
```
{
  "amount": "1234",
  "sigFigs": 2
}
```

is equivalent to "12.34" (amount * 10^-sigFigs)

We must add conversions between FixedPoint and the native serialized format for the blocchains. Also add clone

Helper functions to convert to/from UI "12.34" <-> FixedPoint and functions to add/subtract/compare two FixedPoints can be implemented in client library (wallet-demo) and use native BigInt.

**Comment by webmaster128** [comment]
This can be decomposed into the old form using string operations:

```js
const coin = { amount: "1234", sigFigs: 2, }
const fractional = BigInt(coin.amount.slice(-coin.sigFigs)); // 34n
const whole = BigInt(coin.amount.slice(0, -coin.sigFigs)); // 12n

const coin2 = { amount: "1234", sigFigs: 8, }
const fractional2 = BigInt(coin2.amount.slice(-coin2.sigFigs)); // 1234n
const whole2 = BigInt(coin2.amount.slice(0, -coin2.sigFigs)); // 0n
```

BigInt could be replaced by Long or BN. However, we should give the TypeScript implementation a try that was merged yesterday: https://github.com/Microsoft/TypeScript/pull/25886

There is no native BigDecimal interface yet

> BigInt could form the basis of an eventual BigDecimal implementation.

https://developers.google.com/web/updates/2018/05/bigint

**Comment by webmaster128** [comment]
`amount` -> `quantity` as amount described the whole object.

```ts
interface BcpCoin {
    readonly quantity: string;
    readonly fractionalDigits: number;
}
```

As discussed, BcpCoin could be renamed to BcpAmount.

**Comment by ethanfrey** [comment]
I agree with the format. Let's just look at the three affected types.

Currently `iov-bcp-types/src/bcp.ts`:

```ts
export interface BcpTicker {
  readonly tokenTicker: TokenTicker;
  readonly sigFigs: number;
  readonly tokenName: string;
}


export interface BcpCoin extends BcpTicker {
  readonly whole: number;
  readonly fractional: number;
}
```

And in `transactions.ts`:

```
export interface FungibleToken {
  readonly whole: number;
  readonly fractional: number;
  readonly tokenTicker: TokenTicker;
}
```

------

Changes:

`FungibleToken` -> `BcpAmount` as mentioned above:

```
export interface BcpAmount {
  readonly quantity: string;
  readonly fractionalDigits: number;
  readonly tokenTicker: TokenTicker;
}
```

`type BcpCoin = BcpAmount & BcpTicker` ???
So it has the full value information, but also enhanced with ` tokenName` and `sigFigs` (which is the maximum fractional digits supported by the ticker (maybe change this name).

`BcpCoin` is supposed to have all this info for display, but we can reconsider what exactly belongs inside

**Comment by webmaster128** [comment]
Oh, I was not aware what BcpCoin contains. `FungibleToken` was already renamed to `Amount`. So

1. Use `quantity: string` and `fractionalDigits: number` in `Amount`
2. `type BcpCoin = Amount & BcpTicker` if that works. Otherwise extend `BcpTicker` compatible to 1.
3. Rename `Amount` -> `BcpAmount`?

Regarding 3.: how do we decide what a Bcp* type is?

**Comment by ethanfrey** [comment]
1. and 2. sound good, and I trust you to figure the best way to combine them as a BcpCoin

Re 3. Amount is fine with me. Actually, I don't know if we need the BcpPrefix, if that is also in the package, but maybe it is clearer.

Just do what you think is best with the naming. I just wanted to point out the three different types we use.
